### PR TITLE
Classes and Types for diff tool

### PR DESCRIPTION
**Pull request by unandyala**
Added classes and types for the diff tool. This is the first part of the diff tool refactoring

Added **changedProperty** to the rule definition to capture the changed values (old and new)

1. amfGraphDeltaTypes.ts -  Types describing the AMF graph deltas reported by jsondiffpatch 
2. amfGraphTypes.ts - Types for flattened AMF graph 
3. apiChanges.ts - Class to hold changes between two API specs
4. nodeChanges.ts - Class to hold all changes to a node
5. categorizedChanges.ts - Class to hold categorized changes (when the rule is applied/passed)
6. apiCollectionChanges.ts - Class to hold changes for API collection(to use for directory diff)
7. ruleSet.ts - Class to hold diff rules

### CODE REVIEW

File: src/diff/categorizedChange.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  public ruleName: string;
+
+  //Event type defined in the rule
+  public ruleEvent: string;
+
+  //Category defined in the rule. Example: Breaking, Non-Breaking, Ignore
+  public category: RuleCategory;
+
+  //Flag indicating if a change is ignored by the rule
+  public ignored?: boolean;
```
Comment (wjhsf) [inline src/diff/categorizedChange.ts:23]: This should be easy enough to make a required property, which will make the interface more consistent.

File: src/diff/categorizedChange.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+  //actual value that is changed - [oldValue, newValue]; optional if the rule is ignored
+  public change?: [unknown, unknown];
```
Comment (wjhsf) [inline src/diff/categorizedChange.ts:26]: Conditional types make it possible for this property to be required or excluded based on the value of ignored.

File: src/diff/nodeChanges.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import * as AmfGraphTypes from "./amfGraphTypes";
+import { CategorizedChange } from "./categorizedChange";
+import { RuleCategory } from "./ruleSet";
+
+/**
+ * Class to hold differences of a JSON node
+ */
+export class NodeChanges {
+  public added?: { [key: string]: unknown } | AmfGraphTypes.Node;
+  public removed?: { [key: string]: unknown } | AmfGraphTypes.Node;
```
[the remark this answers is not in the document]
Comment (unandyala) [reply src/diff/nodeChanges.ts:17]: added and removed don't need to have id and type but AmfGraphTypes.Node must have those. Also I want to keep both types (added/removed & Node) separate because they are different

File: src/diff/nodeChanges.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  //categorized changes of the node
+  public categorizedChanges?: CategorizedChange[];
```
[the remark this answers is not in the document]
Comment (unandyala) [reply src/diff/nodeChanges.ts:19]: I was debating on that since they are really optional

File: src/diff/ruleSet.test.ts [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    const rules = {};
+    fs.writeJSONSync(tmpFile.name, JSON.stringify(rules));
+    expect(() => new RuleSet(tmpFile.name)).to.throw(
+      "Rules must be defined as a json array"
+    );
+  });
+
+  it("throws error when the rule has empty conditions", async () => {
+    const tmpFile = tmp.fileSync({ postfix: ".json" });
+    const ruleJson = JSON.parse(buildDefaultRule().toJSON());
```
[the remark this answers is not in the document]
Comment (unandyala) [reply src/diff/ruleSet.test.ts:86]: .toJSON(false) returns RuleSerializable which do not allow an empty conditions object

File: src/diff/nodeChanges.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (wjhsf) [reply src/diff/nodeChanges.ts:17]:
```suggestion
  public added: { [key: string]: unknown };
  public removed: { [key: string]: unknown };
```
I'm confused why you are including `Node`, if you want to keep the separate. And because you're excluding every property defined on `Node` in these type definitions, so there doesn't seem to be a need for it.

File: src/diff/ruleSet.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+export enum RuleCategory {
+  BREAKING = "Breaking",
+  NON_BREAKING = "Non-Breaking",
+  IGNORED = "Ignored"
+}
+
+/**
+ * Holds rules
+ */
+export class RuleSet {
```
[the remark this answers is not in the document]
Comment (unandyala) [reply src/diff/ruleSet.ts:22]: Inheritance makes things complicated for future extensibility. I prefer to leave it as it is for now

File: src/diff/apiCollectionChanges.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+/**
+ * Hold the differencing result of API collections
+ */
+export class ApiCollectionChanges {
+  //Map of api name to its changes
+  changes: Map<string, ApiChanges>;
+
+  //Map of api name to its error message
+  failed: Map<string, string>;
```
[the remark this answers is not in the document]
Comment (unandyala) [reply src/diff/apiCollectionChanges.ts:17]: I will change to failedApis to be consistent with addedApis and removedApi. And change "hasFailures()" to "hasFailedApis()"

File: src/diff/nodeChanges.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (wjhsf) [reply src/diff/nodeChanges.ts:19]: Is there any meaningful difference between the property being unset and the property being set to an empty array?

File: src/diff/ruleSet.test.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (wjhsf) [reply src/diff/ruleSet.test.ts:86]:
```suggestion
    const ruleJson = buildDefaultRule().toJSON(false);
    ruleJson.conditions = {} as TopLevelCondition; // import from json-rules-engine
```
A type assertion avoids that issue.

File: src/diff/nodeChanges.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (unandyala) [reply src/diff/nodeChanges.ts:17]: if the a new node is added or an existing node is removed then I assign the node which is of type AmfGraphTypes.Node to added/removed

File: src/diff/nodeChanges.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (unandyala) [reply src/diff/nodeChanges.ts:19]: Not really. I don't think we want to have that difference.

It depends on how we want our interface to look for the user - undefined or empty array. The general approach seems to be undefined (based on the other code that I have seen). Since we have helper functions I think its fine to have as an empty array

File: src/diff/ruleSet.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (wjhsf) [reply src/diff/ruleSet.ts:22]: A `RuleSet` is fundamentally just a list of rules. I don't think there is any way to extend on that concept that would not be in line with having an `Array` base class. Do you have any specific examples where extending `Array` makes future extensibility more complicated?

File: src/diff/apiCollectionChanges.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (wjhsf) [reply src/diff/apiCollectionChanges.ts:17]: I think that having `Apis` in the property names is redundant. The fact that they are APIs is implied by the name of the class. I think that something like `changed` / `errored` / `added` / `removed` is cleaner than `changedApis` / `failedApis` / `addedApis` / `removedApis`.

File: src/diff/ruleSet.ts [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import { Rule } from "json-rules-engine";
+import fs from "fs-extra";
+
+/**
+ * Category to define in rule
+ */
+export enum RuleCategory {
```
[the remark this answers is not in the document]
Comment (unandyala) [reply src/diff/ruleSet.ts:13]: RuleSet.RuleCategory - not sure. Rule.RuleCategory would have been more appropriate. I will leave it separate

File: src/diff/ruleSet.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (unandyala) [reply src/diff/ruleSet.ts:22]: We may change rules type from array to something else or add additional properties on Ruleset. We don't want to carry all the array properties with Ruleset

File: src/diff/ruleSet.test.ts [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+    rule.event.params.changedProperty = "";
+    const tmpFile = tmp.fileSync({ postfix: ".json" });
+    fs.writeJSONSync(tmpFile.name, [rule.toJSON(false)]);
+
+    expect(() => new RuleSet(tmpFile.name)).to.throw(
+      `Changed property is required in rule: ${rule.name}`
+    );
+  });
+
+  it("throws error when changedProperty is just a space", async () => {
```
[the remark this answers is not in the document]
Comment (unandyala) [reply src/diff/ruleSet.test.ts:171]: Since rules are defined in json file. Felt its good to check for unintentional typo

File: src/diff/categorizedChange.ts (hunk 1 above)
Comment (unandyala) [reply src/diff/categorizedChange.ts:23]: Ignored is not required, because we have RuleCategory for that. I will remove it

File: src/diff/apiCollectionChanges.ts [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import { ApiChanges } from "./apiChanges";
+
+/**
+ * Hold the differencing result of API collections
+ */
+export class ApiCollectionChanges {
```
[the remark this answers is not in the document]
Comment (unandyala) [reply src/diff/apiCollectionChanges.ts:12]: Nothing is really mandatory in this

**Comment by unandyala** [comment]
> Noticed that all the test files have `expect(value).to.equal(true)` (or `false`). These should be changed to `expect(value).to.be.true` (or `false`).
> 
> Additional thoughts on file structure:
> 
> 1. I think it would be cleaner to just have one `types.ts` file.
> 2. There are a lot of new `Changes` files. Maybe put them in a `changes` subdirectory?
> 
> Requesting changes specifically for the classes that have required properties but no constructor. Those classes violate their type definitions.

AmfGraphTypes are generic and can be used by any feature that uses AMF graphs. Delta types are very specific to jsondiffpatch. This is the reason I kept them separate

Rest of the comments are taken care

File: src/diff/apiCollectionChanges.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (unandyala) [reply src/diff/apiCollectionChanges.ts]: I am thinking of adding path to api-config files (both base and new) which will be part of constructor. I will add that when I tie this in the directory diff

File: src/diff/ruleSet.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (wjhsf) [reply src/diff/ruleSet.ts:22]: > We may change rules type from array to something else

What else could it be, though? It's a list loaded from JSON. The only way to represent a list using JSON is with an Array.

> or add additional properties on Ruleset.

Extending Array doesn't prevent that.

> We don't want to carry all the array properties with Ruleset.

Why not? We use a number of them, and there's no harm in leaving the rest exposed.

> In general composition is preferred over inheritance

By creating classes, we've already chosen to use inheritance over composition.

File: src/diff/nodeChanges.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (wjhsf) [reply src/diff/nodeChanges.ts]: I'm not sure I fully understand. Are you sating `nodeChanges.added` could be either of type `Node` or it could be an object with unknown/arbitrary properties?

File: src/diff/apiCollectionChanges.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (wjhsf) [reply src/diff/apiCollectionChanges.ts]: Will those changes be added to this PR?

File: src/diff/ruleSet.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (unandyala) [reply src/diff/ruleSet.ts:22]: This is just making flexible for future which I do not know right now. Inheritance is just one of the features of classes/oop. I feel it has more disadvantages than advantages

File: src/diff/nodeChanges.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (unandyala) [reply src/diff/nodeChanges.ts]: yes. I can exclude id and type as these are duplicated

File: src/diff/apiCollectionChanges.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (unandyala) [reply src/diff/apiCollectionChanges.ts]: I thought of adding in the following.. but I can add now

File: src/diff/ruleSet.ts [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+  BREAKING = "Breaking",
+  NON_BREAKING = "Non-Breaking",
+  IGNORED = "Ignored"
+}
+
+/**
+ * Holds rules
+ */
+export class RuleSet {
```
Comment (wjhsf) [inline src/diff/ruleSet.ts:22]:
```typescript
class RuleSet {
    constructor(public rules: Rule[]) { }
    static async init(rulesPath: string): Promise<RuleSet> {
        return new RuleSet(await fs.readJson(rulesPath));
    }
}
```
The other thread about this suggestion ended up focused on whether to extend Array, but the main point of the suggestion was to use this pattern. Like the `Api` classes, reading from a file should be done in the static method, so that the constructor only has to be concerned with the actual data.

File: src/diff/ruleSet.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (wjhsf) [reply src/diff/ruleSet.test.ts:181]: I don't think that we should do this. There are too many possible typos, so we shouldn't accept responsibility for any of them. (Or, at the very least, sanitizing typos should be separate from rule validation.)

File: src/diff/amfGraphDeltaTypes.ts [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+/**
+ * Type describing the graph delta reported by jsondiffpatch
+ */
+export type GraphDelta = {
+  /**
+   * key: index in the original amf graph (left graph for removed nodes and right graph for added nodes)
+   * value:  NodeDelta - indicate the properties with in a node are changes
+   *        AmfGraphTypes.Node[]  - indicate a node is added/removed/moved
+   */
+  [key: string]: NodeDelta | AmfGraphTypes.Node[];
```
Comment (wjhsf) [inline src/diff/amfGraphDeltaTypes.ts:64]:
```suggestion
type AddedNodes = AmfGraphTypes.Node[];
type RemovedNodes = AmfGraphTypes.Node[];
export type GraphDelta = {
  [key: string]: NodeDelta | AmfGraphTypes.Node[];
```

Minor change, but I think that this would improve readability when looking at type definitions.

Comment (unandyala) [reply src/diff/amfGraphDeltaTypes.ts:64]: not sure.. this feels like added and removed could potentially be different

File: src/diff/ruleSet.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (unandyala) [reply src/diff/ruleSet.test.ts:188]: this is just basic.. it won't hurt to have

File: src/diff/amfGraphDeltaTypes.ts (hunk 12 above)
Comment (wjhsf) [reply src/diff/amfGraphDeltaTypes.ts:64]: They are different, though, even if it's just conceptually. I'm not too concerned about the name / implementation, though. I just think that the type definition should make it easy to understand that a `GraphDelta` represents changed / added / removed nodes. (Right now it reads to me as "change to a node or a list of nodes", and it's not immediately clear what the list is supposed to be.)

File: src/diff/amfGraphDeltaTypes.ts [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import * as AmfGraphTypes from "./amfGraphTypes";
+
+/**
+ * External property, added by jsondiffpath to indicate that the node property is an array
+ */
+export const JSON_DIFF_PATCH_ARRAY_KEY = "_t";
+export const JSON_DIFF_PATCH_ARRAY_VALUE = "a";
+
```
Comment (wjhsf) [inline src/diff/amfGraphDeltaTypes.ts:14]:
```suggestion
export type ArrayDiffMarker = { [JSON_DIFF_PATCH_ARRAY_KEY]: typeof JSON_DIFF_PATCH_ARRAY_VALUE };
```
This object is reused a few times, giving it a name makes it more clear what it's there for.

File: src/diff/amfGraphDeltaTypes.ts [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+/**
+ * Type describing the delta of a node property which is an array
+ */
+export type ArrayDelta = {
+  //key is the index in original array
+  [key: string]: unknown[];
+} & {
+  //"special property to indicate its an array
+  [JSON_DIFF_PATCH_ARRAY_KEY]: string;
+};
```
Comment (wjhsf) [inline src/diff/amfGraphDeltaTypes.ts:32]:
```suggestion
} & ArrayDiffMarker;
```
See comment above. If you don't like this, I'd strongly encourage doing the below change.
```suggestion
} & {
  //"special property to indicate its an array
  [JSON_DIFF_PATCH_ARRAY_KEY]: typeof JSON_DIFF_PATCH_ARRAY_VALUE;
};
```
The value will only ever be `"a"`, so that should be the type, not `string`.

File: src/diff/amfGraphDeltaTypes.ts [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+export type GraphDelta = {
+  /**
+   * key: index in the original amf graph (left graph for removed nodes and right graph for added nodes)
+   * value:  NodeDelta - indicate the properties with in a node are changes
+   *        AmfGraphTypes.Node[]  - indicate a node is added/removed/moved
+   */
+  [key: string]: NodeDelta | AmfGraphTypes.Node[];
+} & {
+  [JSON_DIFF_PATCH_ARRAY_KEY]: string;
+};
```
Comment (wjhsf) [inline src/diff/amfGraphDeltaTypes.ts:67]:
```suggestion
} & ArrayDiffMarker;
```
See comment above. If you don't like this, I'd strongly encourage doing the below change.
```suggestion
} & {
  //"special property to indicate its an array
  [JSON_DIFF_PATCH_ARRAY_KEY]: typeof JSON_DIFF_PATCH_ARRAY_VALUE;
};
```
The value will only ever be `"a"`, so that should be the type, not `string`.

File: src/diff/changes/apiChanges.ts [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+  //Changes between base and new API specifications
+  public nodeChanges: NodeChanges[];
+
+  /**
+   * Create ApiChanges object
+   * @param baseApiSpec - Base API specification to compare
+   * @param newApiSpec - New API specification to compare
+   */
+  constructor(public baseApiSpec: string, public newApiSpec: string) {
+    this.nodeChanges = [];
```
Comment (wjhsf) [inline src/diff/changes/apiChanges.ts:23]: This was previously passed as an argument. Why was that changed?

File: src/diff/changes/apiChanges.ts [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+export class ApiChanges {
+  //Changes between base and new API specifications
+  public nodeChanges: NodeChanges[];
+
+  /**
+   * Create ApiChanges object
+   * @param baseApiSpec - Base API specification to compare
+   * @param newApiSpec - New API specification to compare
+   */
+  constructor(public baseApiSpec: string, public newApiSpec: string) {
```
Comment (wjhsf) [inline src/diff/changes/apiChanges.ts:22]: Are these variables the file paths?

File: src/diff/changes/apiCollectionChanges.ts [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+
+  //array of added apis
+  added: string[];
+
+  /**
+   * Create object to hold changes to two api collections
+   * @param baseApiConfig - Base API config file
+   * @param newApiConfig - New API config file
+   */
+  constructor(public baseApiConfig: string, public newApiConfig: string) {}
```
Comment (wjhsf) [inline src/diff/changes/apiCollectionChanges.ts:30]: Are these variables file paths?

File: src/diff/changes/categorizedChange.ts [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+   * @param category - Category defined in the rule. Example: Breaking, Non-Breaking, Ignore
+   * @param ignored - Flag indicating if a change is ignored by the rule
+   * @param change - actual value that is changed - [oldValue, newValue]; optional if the rule is ignored
+   */
+  constructor(
+    public ruleName: string,
+    public ruleEvent: string,
+    public category: RuleCategory,
+    public change?: [unknown, unknown]
+  ) {
```
Comment (wjhsf) [inline src/diff/changes/categorizedChange.ts:25]:
```suggestion
export class CategorizedChange<C extends RuleCategory> {
  /**
   * @param ruleName - Name of the rule that is applied/passed
   * @param ruleEvent - Event type defined in the rule
   * @param category - Category defined in the rule. Example: Breaking, Non-Breaking, Ignore
   * @param ignored - Flag indicating if a change is ignored by the rule
   * @param change - actual value that is changed - [oldValue, newValue]; optional if the rule is ignored
   */
  constructor(
    public ruleName: string,
    public ruleEvent: string,
    public category: RuleCategory,
    public change?: C extends RuleCategory.IGNORED ? undefined : [unknown, unknown]
  ) {
```
This will ensure that TypeScript reports an error if a change is passed along with category `IGNORED`.

File: src/diff/amfGraphDeltaTypes.ts (hunk 12 above)
Comment (unandyala) [reply src/diff/amfGraphDeltaTypes.ts:64]: type is always same(Node) just the meaning is different. How about type AddedOrRemovedNodes = AmfGraphTypes.Node[];?

File: src/diff/changes/categorizedChange.ts [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import { RuleCategory } from "../ruleSet";
+
+/**
+ * Holds node change that is categorized by rules
+ */
+export class CategorizedChange {
```
Comment (wjhsf) [inline src/diff/changes/categorizedChange.ts:12]: We don't have an `UncategorizedChange` class (at lower levels than this class, we use "delta" instead). Maybe just call this `Change` or `BaseChange`? Either would be a lot shorter.

File: src/diff/amfGraphDeltaTypes.ts (hunk 12 above)
Comment (wjhsf) [reply src/diff/amfGraphDeltaTypes.ts:64]: `AddedOrRemovedNodes` could imply that the things in a single array could be added or removed. Something like `AddedNodesOrRemovedNodes` would be more accurate, but more verbose. (This is why I originally suggested two types, because I couldn't think of a good name for just one.)

File: src/diff/changes/categorizedChange.ts (hunk 19 above)
Comment (unandyala) [reply src/diff/changes/categorizedChange.ts:25]: change is allowed with ignored, its just optional because its hard to determine the change

File: src/diff/changes/categorizedChange.ts (hunk 20 above)
Comment (unandyala) [reply src/diff/changes/categorizedChange.ts:12]: This is change after rule is applied which is categorized- We decided to call it as CategorizedChange, the other option was ClassifiedChange

File: src/diff/amfGraphDeltaTypes.ts [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+export const JSON_DIFF_PATCH_ARRAY_KEY = "_t";
+export const JSON_DIFF_PATCH_ARRAY_VALUE = "a";
+
+/**
+ * Type describing the delta of a node property
+ *
+ * Delta is reported in array format
+ * example: [oldValue, newValue, magicalNumber], magicalNumber is used identify removal (0), text diffs(2), movements(3)
+ */
+export type PropertyDelta = unknown[];
```
[the remark this answers is not in the document]
Comment (unandyala) [reply src/diff/amfGraphDeltaTypes.ts:21]: Both does not look sufficient. We need to separate index moves and actual value changes

File: src/diff/amfGraphDeltaTypes.ts (hunk 13 above)
Comment (unandyala) [reply src/diff/amfGraphDeltaTypes.ts:14]: instead of this I used typeof JSON_DIFF_PATCH_ARRAY_VALUE

File: src/diff/changes/apiChanges.ts (hunk 17 above)
Comment (unandyala) [reply src/diff/changes/apiChanges.ts:22]: yes

File: src/diff/changes/apiCollectionChanges.ts (hunk 18 above)
Comment (unandyala) [reply src/diff/changes/apiCollectionChanges.ts:30]: yes

File: src/diff/changes/apiChanges.ts (hunk 16 above)
Comment (unandyala) [reply src/diff/changes/apiChanges.ts:23]: it does have to be a constructor argument. Changes can be set after the object is created

File: src/diff/changes/categorizedChange.ts [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+   * @param ruleEvent - Event type defined in the rule
+   * @param category - Category defined in the rule. Example: Breaking, Non-Breaking, Ignore
+   * @param ignored - Flag indicating if a change is ignored by the rule
+   * @param change - actual value that is changed - [oldValue, newValue]; optional if the rule is ignored
+   */
+  constructor(
+    public ruleName: string,
+    public ruleEvent: string,
+    public category: RuleCategory,
+    public change?: [unknown, unknown]
```
Comment (wjhsf) [inline src/diff/changes/categorizedChange.ts:24]: This should be updated to use `Modified<unknown>`.
